### DISCUSSION TITLE
Two parallel applications on same machine -> two separate streams or single stream & zmq?

### DISCUSSION THREAD

**Discussion by am11y** [oliver-zehentleitner/unicorn-binance-websocket-api#383] [General]
I have two different, parallel applications which both need Kline and User data stream on the same machine. Do you suggest I use:
1) Single subscriptions and distribute data between apps using zmq or similar?
2) Use subscriptions for each of the apps - make duplicate streams?

**Comment by oliver-zehentleitner** [comment]
I can't give you a general answer to that!

If it's Klines for 1-2 markets, I would do it per process. If you are monitoring 1000 markets, I would do it centrally and outsource it to an extra process. Maybe even two of them, then you can restart a receiver process and the other process continues to work in the meantime. 

How time-critical is your bot? How do you deal with gaps in the data, is that a problem for your bot?

**Comment by adamantmm** [comment]
For a small set of klines + one user-data stream I would duplicate per process. Two managers on the same machine is simple to reason about, and Binance will not care about a handful of identical public kline subscriptions.

I only centralize behind ZMQ/redis when:

- symbol count gets large enough that reconnect storms hurt, or
- both apps need the *same* user-data events and you want a single listen-key lifecycle

User-data is the sharp edge: two processes = two listen keys (fine) vs one shared listen key (fragile if either process renews/closes it). Keep user-data owned by one process and fan out locally if both apps need fills/orders.

**Comment by oliver-zehentleitner** [comment]
On Binance, you always get only one shared list key - never multiple different ones... even if you request a new one, you'll get the currently valid one again.

**Reply by adamantmm** [reply]
Good catch — I had assumed requesting a listen key twice gave you two independent ones. If Binance always returns the currently valid key, two processes on the same account share one lifecycle whether they like it or not. Whichever process closes it (or fails to keep it alive) drops the stream for both.

So even for a small kline set I'd keep a single owner for user-data, and only duplicate the public kline subscriptions. Fan-out of fills then has to be local (ZMQ, unix socket, or in-process). Sorry for the earlier two-listen-key advice.

**Reply by oliver-zehentleitner** [reply]
i assumed that too, but if you delete listenskeys when stopping a userdata stream and the other userdata stream stops too because the listenkey gets invalid you learn this - that was my way to recognize that:D

but this does not say that 2 userdata streams are worthless!  but what i did (if needed) was to use two cheap 3 eur vms with different IPs and run the userdata stream almost independent to each other... (other server, other process, other ip, other network/co location)
